### PR TITLE
Add reverse dns records from pods' ip to fqdn for headless services 

### PR DESCRIPTION
**Pull request by harryge00**
Add DNS records from pods' IPs to headless services' FQDNs.  This can be useful when applications use their own pod IPs to communicate, such as applications registered on zookeepers with their hostnames and GlusterFS [https://github.com/kubernetes/kubernetes/issues/33470](https://github.com/kubernetes/kubernetes/issues/33470). 

Implement this by watching endpoints, and add their IPs to reverseRecordMap.  A map `fqdn2IPsMap` is used to increase time efficiency.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34700)

### CODE REVIEW

**Comment by harryge00** [comment]
cc

**Comment by MrHohn** [comment]
cc @bowei @thockin

**Comment by thockin** [comment]
Can you include more detail in teh PR description, including an example?

**Comment by thockin** [comment]
We already do reverse lookups for pods that elect a subdomain.  This is required because a pod can be "in" multiple headless Services at the same time, but it can only have a single reverse record.  Unless I misunderstand, I don't think we can do this PR

**Comment by harryge00** [comment]
@thockin 
> We already do reverse lookups for pods that elect a subdomain. 

We have just tried deploying a pod with hostname and subdomain:
```
apiVersion: v1
kind: Pod
metadata:
  name: busybox
  namespace: default
spec:
  hostname: busybox-1
  subdomain: default
  containers:
  - image: busybox
    command:
      - sleep
      - "3600"
    name: busybox
```
But we found both the fqdn `busybox-1.default.default.svc.cluster.local` and hostname: `busybox-1` cannot be recognized by other pods. How can we do reverse lookups?

> This is required because a pod can be "in" multiple headless Services at the same time

Yes. We did not take multiple headless services to a single pod into consideration...

**Comment by thockin** [comment]
You need to also define a headless Service that covers your pods, with the
same name as the subdomain.

File: pkg/dns/dns.go [hunk 1]
```diff
@@ -222,6 +226,20 @@ func (kd *KubeDNS) setEndpointsStore() {
 		kcache.ResourceEventHandlerFuncs{
 			AddFunc: kd.handleEndpointAdd,
 			UpdateFunc: func(oldObj, newObj interface{}) {
+				if oldEndPoint, ok := oldObj.(*kapi.Endpoints); ok {
```
Comment (brendandburns) [inline pkg/dns/dns.go]: nit: My preference would be for:
```go
switch (oldEndpoint.(type) {
case *kapi.Endpoints:
  ...
}
```

**Comment by thockin** [comment]
I don't think what you want is feasible.  Closing this for the sake of the TODO list :)

**Pull request closed without merging**